src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -376,53 +349,31 @@ def form(self, form):
                 TypeError,
                 f"Symbolic Hamiltonian should be a ``sympy`` expression but is {type(form)}.",
             )
-        # Calculate number of qubits in the system described by the given
-        # Hamiltonian formula
-        nqubits = 0
-        for symbol in form.free_symbols:
-            if isinstance(symbol, self._qiboSymbol):
-                q = symbol.target_qubit
-            elif isinstance(symbol, sympy.Expr):
-                if symbol not in self.symbol_map:
-                    raise_error(ValueError, f"Symbol {symbol} is not in symbol map.")
-                q, matrix = self.symbol_map.get(symbol)
-                if not isinstance(matrix, self.backend.tensor_types):
-                    # ignore symbols that do not correspond to quantum operators
-                    # for example parameters in the MaxCut Hamiltonian
-                    q = 0
-            if q > nqubits:
-                nqubits = q
-
         self._form = form
-        self.nqubits = nqubits + 1
+        self.nqubits = self._calculate_nqubits_from_form(form)
 
-    @property
+    @cached_property
     def terms(self):
         """List of terms of which the Hamiltonian is a sum of.
 
         Terms will be objects of type :class:`qibo.core.terms.HamiltonianTerm`.
         """
-        if self._terms is None:
-            # Calculate terms based on ``self.form``
-            from qibo.hamiltonians.terms import (  # pylint: disable=import-outside-toplevel
-                SymbolicTerm,
-            )
+        # Calculate terms based on ``self.form``
+        from qibo.hamiltonians.terms import (  # pylint: disable=import-outside-toplevel
+            SymbolicTerm,
+        )
```

Comment:
I think we could try to move this top level as well. `terms` does not seem to import anything from `hamiltonians`, so if there are circular import issues we may be able to resolve them relatively easily.